services/venice-controller/src/main/java/com/linkedin/venice/controller/DeferredVersionSwapService.java

```diff
@@ -1094,6 +892,240 @@ private Runnable getRunnableForDeferredVersionSwap() {
     };
   }
 
+  private void performSequentialRollForward(
+      String cluster,
+      List<Store> parentStores,
+      Map<String, ControllerClient> childControllerClientMap,
+      List<String> rolloutOrder) {
+    for (Store parentStore: parentStores) {
+      int targetVersionNum = parentStore.getLargestUsedVersionNumber();
+      if (targetVersionNum < 1) {
+        continue;
+      }
+
+      Version targetVersion = parentStore.getVersion(targetVersionNum);
+      if (targetVersion == null) {
+        String message =
+            "Parent version is null for store " + parentStore.getName() + " for target version " + targetVersionNum;
+        logMessageIfNotRedundant(message);
+        continue;
+      }
+
+      if (!isTargetRegionPushWithDeferredSwapEnabled(targetVersion)) {
+        continue;
+      }
+
+      String storeName = parentStore.getName();
+      Set<String> targetRegions = Collections.singleton(rolloutOrder.get(0));
+      Set<String> remainingRegions = getRegionsForVersionSwap(childControllerClientMap, targetRegions);
+
+      // Check if the target version is in a terminal state (push job completed or failed)
+      long startTime = System.currentTimeMillis();
+      if (!isPushInTerminalState(
+          targetVersion,
+          cluster,
+          parentStore.getName(),
+          targetVersionNum,
+          remainingRegions,
+          targetRegions)) {
+        logLatency(startTime, storeName, targetVersionNum);
+        continue;
+      }
+
+      // Check if the cached waitTime for the target version has elapsed
+      String kafkaTopicName = Version.composeKafkaTopic(storeName, targetVersionNum);
+      Long cachedWaitTime = storeWaitTimeCacheForSequentialRollout.getIfPresent(kafkaTopicName);
+      if (cachedWaitTime != null) {
+        if (!didWaitTimePassInRegionForSequentialRollout(parentStore, cachedWaitTime)) {
```

Comment:
cannot comment on the previous code, nit: can we rename one of these method, may be the one which also might add to cache as `isRegionReadyForRollout`